benchmark/runner/mini_runners.cpp

```diff
@@ -0,0 +1,889 @@
+    brain::ExecutionOperatingUnitFeatureVector pipe0_vec;
+    brain::ExecutionOperatingUnitFeatureVector pipe1_vec;
+    pipe0_vec.emplace_back(brain::ExecutionOperatingUnitType::SEQ_SCAN, row, 4 * num_col, num_col, car);
+    pipe0_vec.emplace_back(brain::ExecutionOperatingUnitType::SORT_BUILD, row, 4 * num_col, num_col, car);
+    pipe1_vec.emplace_back(brain::ExecutionOperatingUnitType::SORT_ITERATE, row, 4 * num_col, num_col, car);
+    units.RecordOperatingUnit(execution::pipeline_id_t(0), std::move(pipe0_vec));
+    units.RecordOperatingUnit(execution::pipeline_id_t(1), std::move(pipe1_vec));
+
+    std::stringstream cols;
+    for (auto i = 1; i <= num_col; i++) {
+      cols << "col" << i;
+      if (i != num_col) {
+        cols << ", ";
+      }
+    }
+
+    std::stringstream query;
+    query << "SELECT " << (cols.str()) << " FROM "
+          << execution::sql::TableGenerator::GenerateTableName(type::TypeId::INTEGER, 15, row, car) << " ORDER BY "
+          << (cols.str());
+    BenchmarkSqlStatement(query.str(), &units, std::make_unique<optimizer::TrivialCostModel>(), true);
+    metrics_manager_->Aggregate();
+    metrics_manager_->UnregisterThread();
+  }
+
+  state.SetItemsProcessed(row);
+}
+
+/**
+ * Arg: <0, 1, 2>
+ * 0 - Number of columns to select
+ * 1 - Number of rows
+ * 2 - Cardinality
+ */
+BENCHMARK_REGISTER_F(MiniRunners, SortRunners)->Unit(benchmark::kMillisecond)->Iterations(1)->Apply(GenScanArguments);
+
+// NOLINTNEXTLINE
+BENCHMARK_DEFINE_F(MiniRunners, HashJoinRunners)(benchmark::State &state) {
+  auto num_col = state.range(0);
+  auto row = state.range(1);
+  auto car = state.range(2);
+
+  // NOLINTNEXTLINE
+  for (auto _ : state) {
+    metrics_manager_->RegisterThread();
+
+    brain::PipelineOperatingUnits units;
+    brain::ExecutionOperatingUnitFeatureVector pipe0_vec;
+    brain::ExecutionOperatingUnitFeatureVector pipe1_vec;
+    pipe0_vec.emplace_back(brain::ExecutionOperatingUnitType::SEQ_SCAN, row, 4 * num_col, num_col, car);
+    pipe0_vec.emplace_back(brain::ExecutionOperatingUnitType::HASHJOIN_BUILD, row, 4 * num_col, num_col, car);
+    pipe1_vec.emplace_back(brain::ExecutionOperatingUnitType::SEQ_SCAN, row, 4 * num_col, num_col, car);
+    pipe1_vec.emplace_back(brain::ExecutionOperatingUnitType::HASHJOIN_PROBE, row, 4 * num_col, num_col,
+                           row * row / car);
+    pipe1_vec.emplace_back(brain::ExecutionOperatingUnitType::OP_INTEGER_COMPARE, row, 4, 1, row * row / car);
+    units.RecordOperatingUnit(execution::pipeline_id_t(0), std::move(pipe0_vec));
+    units.RecordOperatingUnit(execution::pipeline_id_t(1), std::move(pipe1_vec));
+
+    std::stringstream query;
+    auto tbl_name = execution::sql::TableGenerator::GenerateTableName(type::TypeId::INTEGER, 15, row, car);
+    query << "SELECT b.col15 FROM " << tbl_name << ", " << tbl_name << " as b WHERE ";
+    for (auto i = 1; i <= num_col; i++) {
+      query << tbl_name << ".col15 = b.col15";
```

Comment:
Ah! I was actually thinking about duplicating the columns when generating the tables in `table_generator.cpp`. Is that doable/easy to do? Because I'm a bit concerned that the optimizer may simplify the duplicated join predicate and we only end up with joining on one column. Even though our optimize may not optimize this now, later modifications (like the new rewriting rules from Erik) may just change the behavior.

Reply:
Tweaked. `table_generator.cpp` now generates col16-col31 which duplicate the values of col15. The join predicate for hash join and nested loop join would then use col15 through col31.